types/address/hash.go

```diff
@@ -59,13 +63,28 @@ func Compose(typ string, subAddresses []Addressable) ([]byte, error) {
 }
 
 // Module is a specialized version of a composed address for modules. Each module account
-// is constructed from a module name and module account key.
-func Module(moduleName string, key []byte) []byte {
-	mKey := append([]byte(moduleName), 0)
-	return Hash("module", append(mKey, key...))
+// is constructed from a module name and a sequence of derivation keys (at least one
+// derivation key must be provided). The derivation keys must be unique
+// in the module scope, and is usually constructed from some object id. Example, let's
+// a x/dao module, and a new DAO object, it's address would be:
+//
+//	address.Module(dao.ModuleName, newDAO.ID)
+func Module(moduleName string, derivationKeys ...[]byte) []byte {
+	moduleNameBz := []byte(moduleName)
+	if len(derivationKeys) == 0 {
+		return crypto.AddressHash([]byte(moduleNameBz))
+	}
+	mKey := append(moduleNameBz, 0)
+	addr := Hash("module", append(mKey, derivationKeys[0]...))
+	for _, k := range derivationKeys[1:] {
+		addr = Derive(addr, k)
+	}
```

Comment:
I have 2 questions about this:
1. Why do we need to append `0`?
2. We can simply always use Derive, and no custom Hash.

My proposal, to make it simpler, would be:
```go
// keep the separate `if` case when len(derivationKeys)==0

addr := append([]byte("module"), moduleNameBz)
for _, k := range derivationKeys {
  addr = Derive(addr, k)
}
return addr
```

Reply:
I explained it in the issue: https://github.com/cosmos/cosmos-sdk/issues/13910#issuecomment-1320371999 (and 2 comments below).

The motivation was to have a shortcut, as seen in groups module:
* `address.Derive( address.Module(group.ModuleName, []byte{GroupPolicyTablePrefix}), buf)`
* vs `address.Derive( address.Derive( address.Module(group.ModuleName), []byte{GroupPolicyTablePrefix}), buf)`

Reply:
Appending `0` is an outcome of that decision -- it's needed to safely separate (and avoid clash) the module name and the first derivation key.